util/setuidgid.c

```diff
@@ -0,0 +1,341 @@
+			warnp("Failed to allocate memory");
+			goto err0;
+		}
+		strncpy(*username_p, combined, pos);
+		(*username_p)[pos] = '\0';
+
+		/* Extract groupname. */
+		if ((*groupname_p = malloc((len - pos + 1))) == NULL) {
+			warnp("Failed to allocate memory");
+			goto err1;
+		}
+		strncpy(*groupname_p, combined + pos + 1, len - pos);
+		(*groupname_p)[len-pos] = '\0';
+	} else {
+		/* Duplicate string. */
+		if ((*username_p = strdup(combined)) == NULL)
+			goto err0;
+		*groupname_p = NULL;
+	}
+
+	/* Success! */
+	return (0);
+
+err1:
+	free(*username_p);
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * setuidgid(user_group_string, leave_supplementary):
+ * Set the UID and/or GID to the names given in ${user_group_string}.
+ * Depending on the existence and position of a colon ":", the behaviour is
+ * - no ":" means that the string is a username.
+ * - ":" in the first position means that the string is a groupname.
+ * - otherwise, the string is parsed into "username:grouname".
+ *
+ * The behaviour with supplementary groups depends on ${leave_supplementary}:
+ * - SETUIDGID_SGROUP_IGNORE: do not attempt to leave supplementary groups.
+ * - SETUIDGID_SGROUP_LEAVE_WARN: attempt to leave; if it fails, give a
+ *   warning but continue.
+ * - SETUIDGID_SGROUP_LEAVE_ERROR: attempt to leave; if it fails, return
+ *   an error.
+ */
+int
+setuidgid(const char * opt_usergroup, int leave_supplementary)
+{
+	char * username;
+	char * groupname;
+
+	/* Get the username:groupname from the opt_user string. */
+	if (string_extract_user_group(opt_usergroup, &username, &groupname)
+	    != 0)
+		goto err0;
+
+	/* If requested, leave supplementary groups. */
+	if (leave_supplementary != SETUIDGID_SGROUP_IGNORE) {
+		/* Leave any supplementary groups. */
+		if (setgroups_none() != 0) {
+			/*
+			 * Quit before warnp() to avoid clearing errno.
+			 * Do this regardless of whether the user specified
+			 * SETUIDGID_SGROUP_LEAVE_ERROR or not, because if
+			 * setgroups() bails with EPERM then so will setgid()
+			 * and setuid().
+			 */
+			if (errno == EPERM)
+				goto err1;
+			warnp("setgroups()");
+			if (leave_supplementary == SETUIDGID_SGROUP_LEAVE_ERROR)
+				goto err1;
+		}
+
+		/* Check that we have actually left all supplementary groups. */
+		if (check_supplementary_groups_none() != 0) {
+			warn0("Failed to leave supplementary groups");
+			if (leave_supplementary == SETUIDGID_SGROUP_LEAVE_ERROR)
+				goto err1;
+		}
+	}
+
+	/* Set group; must be changed before user. */
+	if (groupname != NULL) {
+		if (set_group(groupname) != 0) {
+			/* Quit before warnp() to avoid clearing errno. */
+			if (errno == EPERM)
+				goto err1;
+			warn0("Error changing group");
+			goto err1;
+		}
+	}
+
+	/* Set user. */
+	if (username != NULL) {
+		if (set_user(username) != 0) {
+			/* Quit before warnp() to avoid clearing errno. */
+			if (errno == EPERM)
+				goto err1;
+			warn0("Error changing user");
+			goto err1;
```

Comment:
This idiom (occurring many times) could be written

```
if (errno != EPERM)
        warn0("blah blah blah");
goto err1;
```

Seems a bit cleaner to me?